File: MathCore.DSP/Filters/ChebyshevHighPass.cs
using System.ComponentModel;
using System.Diagnostics;

using MathCore.DSP.Infrastructure;

namespace MathCore.DSP.Filters;

/// <summary>Фильтр Чебышева нижних частот</summary>
public class ChebyshevHighPass : ChebyshevFilter
{
    private static (double[] A, double[] B) InitializeI(Specification Spec)
    {
        var N = (int)Math.Ceiling(arcch(Spec.kEps) / arcch(Spec.kW)); // Порядок фильтра
        Debug.Assert(N > 0, $"N > 0 :: {N} > 0");

        var poles = GetNormedPolesI(N, Spec.EpsP);

        var high_pass_poles = TransformToHighPassW(poles, Spec.Ws);

        var z_poles = ToZArray(high_pass_poles, Spec.dt);

        var k_poles = z_poles.Multiply(z => (1 + z) / 2).Re;

        var g_norm = N.IsEven()
            ? Spec.Gp * k_poles
            : 1 * k_poles;

        var B = new double[N + 1];
        for (var i = 0; i < B.Length; i++)
            B[i] = SpecialFunctions.BinomialCoefficient(N, i) * (i % 2 == 0 ? g_norm : -g_norm);
        var A = Polynom.Array.GetCoefficientsInverted(z_poles).ToRe();

        return (A, B);
    }

    private static (double[] A, double[] B) InitializeII(Specification Spec)
    {
        var N = (int)Math.Ceiling(arcch(Spec.kEps) / arcch(Spec.kW)); // Порядок фильтра
        Debug.Assert(N > 0, $"N > 0 :: {N} > 0");
        var (zeros, poles) = GetNormedPolesII(N, Spec.EpsS);

        var high_pass_zeros = TransformToHighPassW(zeros, Spec.Ws);
        if (N.IsOdd())
            high_pass_zeros = high_pass_zeros.Prepend(0);
        var high_pass_poles = TransformToHighPassW(poles, Spec.Ws);

        var z_zeros = ToZArray(high_pass_zeros, Spec.dt);
        var z_poles = ToZArray(high_pass_poles, Spec.dt);

        var B = Polynom.Array.GetCoefficientsInverted(z_zeros).ToRe();
        var A = Polynom.Array.GetCoefficientsInverted(z_poles).ToRe();

        var k_zeros = z_zeros.Multiply(z => 1 + z).Re;
        var k_poles = z_poles.Multiply(z => 1 + z).Re;

        var g_norm = N.IsEven()
            ? Spec.Gp * k_poles / k_zeros
            : 1 * k_poles / k_zeros;

        B.Multiply(g_norm);

        return (A, B);
    }

    private static (double[] A, double[] B) InitializeIICorrected(Specification Spec)
    {
        var N = (int)Math.Ceiling(arcch(Spec.kEps) / arcch(Spec.kW)); // Порядок фильтра
        Debug.Assert(N > 0, $"N > 0 :: {N} > 0");

        var (zeros, poles) = GetNormedPolesII(N, Spec.EpsS, Spec.kw);

        var high_pass_zeros = TransformToHighPassW(zeros, Spec.Ws);
        if (N.IsOdd())
            high_pass_zeros = high_pass_zeros.Prepend(0);
        var high_pass_poles = TransformToHighPassW(poles, Spec.Ws);

        var z_zeros = ToZArray(high_pass_zeros, Spec.dt);
        var z_poles = ToZArray(high_pass_poles, Spec.dt);

        var B = Polynom.Array.GetCoefficientsInverted(z_zeros).ToRe();
        var A = Polynom.Array.GetCoefficientsInverted(z_poles).ToRe();

        var k_zeros = z_zeros.Multiply(z => 1 + z).Re;
        var k_poles = z_poles.Multiply(z => 1 + z).Re;

        var g_norm = N.IsEven()
            ? Spec.Gp * k_poles / k_zeros
            : 1 * k_poles / k_zeros;

        B.Multiply(g_norm);

        return (A, B);
    }

    /// <summary>Инициализация нового фильтра Чебышева нижних частот</summary>
    /// <param name="dt">Период дискретизации</param>
    /// <param name="fs">Частота заграждения</param>
    /// <param name="fp">Частота пропускания</param>
    /// <param name="Gp">Затухание в полосе пропускания (0.891250938 = -1 дБ)</param>
    /// <param name="Gs">Затухание в полосе заграждения (0.01        = -40 дБ)</param>
    /// <param name="Type">Тип (род) фильтра чебышева</param>
    public ChebyshevHighPass(double dt,
        double fs,
        double fp,
        double Gp = 0.891250938,
        double Gs = 0.01,
        ChebyshevType Type = ChebyshevType.I)
        : this(GetSpecification(dt, fs, fp, Gp, Gs), Type) { }

    public ChebyshevHighPass(Specification Spec, ChebyshevType Type = ChebyshevType.I)
        : this(Type switch
        {
            ChebyshevType.I => InitializeI(Spec),
            ChebyshevType.II => InitializeII(Spec),
            ChebyshevType.IICorrected => InitializeIICorrected(Spec),
            _ => throw new InvalidEnumArgumentException(nameof(Type), (int)Type, typeof(ChebyshevType))
        }, Spec, Type) { }

    private ChebyshevHighPass((double[] A, double[] B) config, Specification Spec, ChebyshevType Type) : base(config.B, config.A, Spec, Type) { }
}

File: MathCore.DSP/Filters/Python/filters/chebyshev_highpass.py
# Высокочастотный фильтр Чебышева (порт C# ChebyshevHighPass)
import numpy as np
from math import ceil

from .chebyshev_filter import ChebyshevFilter, ChebyshevType
from .analog_based_filter import AnalogBasedFilter, Specification
from .digital_filter import DigitalFilter


class ChebyshevHighPass(ChebyshevFilter):
    """
    Цифровой высокочастотный фильтр Чебышева (ФВЧ).

    Для LP→HP трансформации используются нормированные полюса ФНЧ Чебышева
    и трансформация p → Ws/p, где Ws — циклическая частота среза.
    """

    @staticmethod
    def _initialize_i(spec: Specification):
        """Коэффициенты ФВЧ Чебышева I рода."""
        N = int(ceil(ChebyshevFilter.arcch(spec.kEps) / ChebyshevFilter.arcch(spec.kW)))

        poles = ChebyshevFilter.get_normed_poles_i(N, spec.EpsP)

        # LP→HP трансформация: p → Ws/p
        hp_poles = AnalogBasedFilter.transform_to_high_pass_w(poles, spec.Ws)

        z_poles = DigitalFilter.to_z_array(hp_poles, spec.dt)

        # нормировка по частоте Найквиста (z = -1)
        k_poles = np.prod([(1.0 + z) / 2.0 for z in z_poles]).real
        g_norm = (spec.Gp if N % 2 == 0 else 1.0) * k_poles

        binomial = AnalogBasedFilter.get_binomial_coefficients(N)
        B = np.array([b * (g_norm if i % 2 == 0 else -g_norm)
                      for i, b in enumerate(binomial)])
        A = np.poly(z_poles).real

        return A, B

    @staticmethod
    def _initialize_ii(spec: Specification):
        """Коэффициенты ФВЧ Чебышева II рода."""
        N = int(ceil(ChebyshevFilter.arcch(spec.kEps) / ChebyshevFilter.arcch(spec.kW)))

        zeros, poles = ChebyshevFilter.get_normed_poles_ii(N, spec.EpsS)

        # LP→HP трансформация нулей и полюсов
        hp_zeros = AnalogBasedFilter.transform_to_high_pass_w(zeros, spec.Ws)
        if N % 2 != 0:
            hp_zeros = [0.0 + 0j] + list(hp_zeros)  # дополнительный нуль в 0
        hp_poles = AnalogBasedFilter.transform_to_high_pass_w(poles, spec.Ws)

        z_zeros = DigitalFilter.to_z_array(hp_zeros, spec.dt)
        z_poles = DigitalFilter.to_z_array(hp_poles, spec.dt)

        B = np.poly(z_zeros).real
        A = np.poly(z_poles).real

        # нормировка по частоте Найквиста
        k_zeros = np.prod([1.0 + z for z in z_zeros]).real
        k_poles = np.prod([1.0 + z for z in z_poles]).real
        g_norm = k_poles / k_zeros

        B = B * g_norm
        return A, B

    @staticmethod
    def _initialize_ii_corrected(spec: Specification):
        """Коэффициенты ФВЧ Чебышева II рода с коррекцией частот."""
        N = int(ceil(ChebyshevFilter.arcch(spec.kEps) / ChebyshevFilter.arcch(spec.kW)))

        # W0 = kw для коррекции прототипа
        zeros, poles = ChebyshevFilter.get_normed_poles_ii(N, spec.EpsS, spec.kw)

        hp_zeros = AnalogBasedFilter.transform_to_high_pass_w(zeros, spec.Ws)
        if N % 2 != 0:
            hp_zeros = [0.0 + 0j] + list(hp_zeros)
        hp_poles = AnalogBasedFilter.transform_to_high_pass_w(poles, spec.Ws)

        z_zeros = DigitalFilter.to_z_array(hp_zeros, spec.dt)
        z_poles = DigitalFilter.to_z_array(hp_poles, spec.dt)

        B = np.poly(z_zeros).real
        A = np.poly(z_poles).real

        k_zeros = np.prod([1.0 + z for z in z_zeros]).real
        k_poles = np.prod([1.0 + z for z in z_poles]).real
        g_norm = k_poles / k_zeros

        B = B * g_norm
        return A, B

    def __init__(self, dt: float, fs: float, fp: float,
                 Gp: float = 0.891250938, Gs: float = 0.01,
                 filter_type: ChebyshevType = ChebyshevType.I):
        """
        Создаёт ФВЧ Чебышева.

        Параметры
        ---------
        dt          : период дискретизации (с)
        fs          : частота заграждения (Гц), fs < fp
        fp          : граничная частота полосы пропускания (Гц)
        Gp          : коэффициент передачи на границе пропускания
        Gs          : коэффициент передачи на границе заграждения
        filter_type : тип фильтра
        """
        spec = AnalogBasedFilter.get_specification(dt, fs, fp, Gp, Gs)
        init = {
            ChebyshevType.I:           ChebyshevHighPass._initialize_i,
            ChebyshevType.II:          ChebyshevHighPass._initialize_ii,
            ChebyshevType.IICorrected: ChebyshevHighPass._initialize_ii_corrected,
        }
        A, B = init[filter_type](spec)
        super().__init__(B, A, spec, filter_type)
